src/uct/ib/base/ib_iface.c

```diff
@@ -597,6 +597,133 @@ static void uct_ib_iface_res_domain_cleanup(uct_ib_iface_res_domain_t *res_domai
 #endif
 }
 
+#if HAVE_DECL_IBV_EXP_QUERY_GID_ATTR
+static size_t uct_iface_get_priority_gid_index(unsigned *gid_idx_rates, int len)
+{
+    unsigned best_priority = gid_idx_rates[0];
+    size_t gid_index       = 0;
+    int i;
+
+    for (i = 1; i < len; i++) {
+        if (gid_idx_rates[i] < best_priority) {
+            best_priority = gid_idx_rates[i];
+            gid_index = i;
+        }
+    }
+
+    return gid_index;
+}
+#endif
+
+static int uct_iface_is_gid_ipv4(union ibv_gid *gid, int gid_index)
+{
+    const struct in6_addr *raw = (struct in6_addr *)gid->raw;
+    char p[128];
+
+    ucs_debug("testing is_ipv4 on gid index %d: %s",
+              gid_index, inet_ntop(AF_INET6, gid, p, sizeof(p)));
+
+    return ((raw->s6_addr32[0] | raw->s6_addr32[1]) |
+            (raw->s6_addr32[2] ^ htonl(0x0000ffff))) == 0UL ||
+           /* IPv4 encoded multicast addresses */
+           (raw->s6_addr32[0] == htonl(0xff0e0000) &&
+            ((raw->s6_addr32[1] | (raw->s6_addr32[2] ^ htonl(0x0000ffff))) == 0UL));
+}
+
+static ucs_status_t uct_ib_iface_set_roce_gid_index(uct_ib_device_t *dev,
+                                                    uint8_t port_num,
+                                                    size_t *gid_index)
+{
+    int i, gid_tbl_len      = uct_ib_device_port_attr(dev, port_num)->gid_tbl_len;
+    ucs_status_t status     = UCS_OK;
+#if HAVE_DECL_IBV_EXP_QUERY_GID_ATTR
+    struct ibv_exp_gid_attr attr;
+    unsigned *gid_idx_rates = ucs_alloca(gid_tbl_len * sizeof(*gid_idx_rates));
+#else
+    union ibv_gid gid;
+#endif
+
+    for (i = 0; i < gid_tbl_len; i++) {
+#if HAVE_DECL_IBV_EXP_QUERY_GID_ATTR
+        attr.comp_mask = IBV_EXP_QUERY_GID_ATTR_TYPE | IBV_EXP_QUERY_GID_ATTR_GID;
+        if (ibv_exp_query_gid_attr(dev->ibv_context, port_num, i, &attr)) {
+            ucs_error("failed to query gid attributes "
+                      "(" UCT_IB_IFACE_FMT" gid_idx %d). %m",
+                      uct_ib_device_name(dev), port_num, i);
+            status = UCS_ERR_INVALID_PARAM;
+            goto out;
+        }
+
+        switch (attr.type) {
+        case IBV_EXP_ROCE_V2_GID_TYPE:
```

Comment:
i prefer we have a static array which defines priorities per gid version/type, and make it configurable later